### PR TITLE
feat(data-events): track content gate interactions

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

1200550061930446-as-1205734635694818

Creates `gate_interaction` data events with the following schema:

| Name           | Type     | Obs                                                                                               |
| -------------- | -------- | ------------------------------------------------------------------------------------------------- |
| `gate_post_id` | `int`    |                                                                                                   |
| `action`       | `string` | Either `seen`, `form_submission_received`, `form_submission_success` or `form_submission_failure` |
| `action_type`  | `string` | Either `paid_membership` or `registration`. Not applicable when `action` is `seen`                |
| `referer`      | `string` |                                                                                                   |
| `order_id`     | `int`    | Only applicable when `action_type` is `paid_membership`                                           |
| `product_id`   | `int`    | Only applicable when `action_type` is `paid_membership`                                           |
| `amount`       | `float`  | Only applicable when `action_type` is `paid_membership`                                           |
| `currency`     | `string` | Only applicable when `action_type` is `paid_membership`                                           |

To allow tracking form interactions inside the gate, a script injects a hidden `memberships_content_gate` input to forms inside the gate. This is passed by the reader registration or the modal checkout logic so it can finally get picked up by the data event listeners.

It also implements the integration with the ga4 connector, which will include additional parameters to the GA event: `logged_in`, `is_reader`, and `email_hash`.

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-blocks/pull/1590
2. Make sure you have GA4 connected with the wp options `ga4_measurement_id` and `ga4_measurement_protocol_secret` configured
3. Set up a membership plan restricted by user registration and configure an inline content gate with a reg. wall (reader registration block)
4. Go through the gate anonymously and make sure your session allows 3rd party cookies
5. Visit GA's realtime dashboard, wait until the events are sent and confirm you get 3 `np_gate_interaction` events: `seen` and `form_submission_received` and `form_submission_success` with `action_type` set to `registration` and the correct `referer`
6. Change the membership plan to grant access through a subscription product
7. Change the content gate from "inline" to the "overlay" style
8. Repeat steps 3 and 4 but confirm that this time the `action_type` is `paid_membership`
9. Confirm the event also includes `order_id`, `product_id`, `amount` and `currency`

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by leogermani — APPROVED** [review]
Everything works great!

One thing I noticed is: After I had 2 successful form submissions (one via registration and one product purchase), I noticed that the `product_id` parameter appears to have two entries with the product ID (where it should be only one)

image


This might be just a glitch of the Realtime Report... I think to confirm it we'll need to look at how the data looks in BigQuery... 

But maybe if we simply didn't sent the parameter when they are empty, instead of sending empry strings? WDYT?

File: includes/reader-activation/class-reader-activation.php [hunk 1]
```diff
@@ -1630,6 +1630,15 @@ public static function register_reader( $email, $display_name = '', $authenticat
...
 		}
 
+		/**
+		 * Filters the metadata to pass along to the action hook.
+		 *
+		 * @param array          $metadata      Metadata.
+		 * @param int|false      $user_id       The created user id or false if the user already exists.
+		 * @param false|\WP_User $existing_user The existing user object.
+		 */
+		$metadata = apply_filters( 'newspack_register_reader_metadata', $metadata, $user_id, $existing_user );
```
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/reader-activation/class-reader-activation.php:1640]: It's not, it seemed like a useful filter to add, given that [this one](https://github.com/Automattic/newspack-plugin/pull/2740/files#diff-5f86ac04d34e252d46cd8e2d5d644a2c10676d3dc7d0a5ef39f2aae08a0f5b7dR438-R444) was also added. Do you think it should be removed?

[the remark this answers is not in the document]
Comment (leogermani) [reply includes/reader-activation/class-reader-activation.php:1640]: no, it's fine

**Comment by miguelpeixe** [comment]
The 2 events with `product_id` from the `form_submission_received` and `form_submission_success` for the same `paid_membership` action type.

**Pull request merged**